internal/controller/controllers/image_controller.go

```diff
@@ -17,29 +17,154 @@ limitations under the License.
 package controllers
 
 import (
-	"github.com/sirupsen/logrus"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
 
+	"github.com/openshift/assisted-service/internal/common"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/openshift/assisted-service/internal/bminventory"
+	adiiov1alpha1 "github.com/openshift/assisted-service/internal/controller/api/v1alpha1"
+	"github.com/openshift/assisted-service/models"
+	"github.com/openshift/assisted-service/restapi/operations/installer"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-
-	adiiov1alpha1 "github.com/openshift/assisted-service/internal/controller/api/v1alpha1"
 )
 
 // ImageReconciler reconciles a Image object
 type ImageReconciler struct {
 	client.Client
-	Log    logrus.FieldLogger
-	Scheme *runtime.Scheme
+	Log       logrus.FieldLogger
+	Scheme    *runtime.Scheme
+	Installer bminventory.InstallerInternals
 }
 
 // +kubebuilder:rbac:groups=adi.io.my.domain,resources=images,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=adi.io.my.domain,resources=images/status,verbs=get;update;patch
 
 func (r *ImageReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	// your logic here
+	ctx := context.Background()
+
+	image := &adiiov1alpha1.Image{}
+	if err := r.Get(ctx, req.NamespacedName, image); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrap(err, "failed to get image data")
+	}
+
+	return r.create(ctx, image)
```

Comment:
What happens if it already exists? Ideally you'd want a function named something like "ensureImage" that makes sure the image exists, but does not promise to carry out any particular imperative action.

Reply:
If image already exists backed will just return, if some parameter have changed it will regenerate it.
We did it because we know how backend is implemented.